app/code/Magento/Catalog/Model/CategoryRepository.php

```diff
@@ -82,6 +82,10 @@ public function save(\Magento\Catalog\Api\Data\CategoryInterface $category)
         $existingData = array_diff_key($existingData, array_flip(['path', 'level', 'parent_id']));
         $existingData['store_id'] = $storeId;
 
+        if ($category->getData('save_rewrites_history', null) !== null) {
```

Comment:
It should be placed in Magento_CatalogUrlRewrite, not in Magento_Catalog.

Reply:
Do you have a clue how we could do that?

Reply:
@swnsma 

In the first place my code is very similar to the solution products, just the categories are persisted differently and I am wondering which is the better way :)
I.e. if we change it for categories, we might need to adapt products as well.

I see several options:

1. Mess with \Magento\Catalog\Model\CategoryRepository::getExtensibleDataObjectConverter (which is already deprecated)
2. Intercept a different method, in the before save event, and somehow get the info that rewrites should be created from the request
3. .... ?

Help is very much appreciated :)

Reply:
The problem that approach is lead to violation of modularity.
I see that for save from Adminhtml area, it uses observer to set a value and ResourceModel to save:
```observer name="category_save_rewrites_history_setter"```

Could you play somehow with `\Magento\Catalog\Model\CategoryRepository::getExtensibleDataObjectConverter` ? (the method is deprecated because class should be injected into the constructor parameters).

I see that it can handle custom attributes: \Magento\Framework\Api\ExtensibleDataObjectConverter::processCustomAttributes
Could we check why it does not help to solve our issue?

Reply:
> The problem that approach is lead to violation of modularity.
absolutely

> I see that for save from Adminhtml area, it uses observer to set a value and ResourceModel to save:
> observer name="category_save_rewrites_history_setter"

yes, but no access to what was sent via the API, here

> Could you play somehow with \Magento\Catalog\Model\CategoryRepository::getExtensibleDataObjectConverter ? (the
> method is deprecated because class should be injected into the constructor parameters).
> I see that it can handle custom attributes:
> \Magento\Framework\Api\ExtensibleDataObjectConverter::processCustomAttributes
> Could we check why it does not help to solve our issue?

Maybe ... I am still wondering if that flag should be a custom attribute at all

Reply:
After thinking a bit about it, I think it should be a custom attribute ... will check the converter

Reply:
Just adding save_rewrites_history to \Magento\Catalog\Model\Category::getCustomAttributesCodes seems to do the trick -- not even messing with InputParamsResolver needed :)

Reply:
but of course 

	    protected function getCustomAttributesCodes()
	    {
		if ($this->customAttributesCodes === null) {
		    $this->customAttributesCodes = $this->getEavAttributesCodes($this->metadataService);
		    $this->customAttributesCodes = array_diff($this->customAttributesCodes, CategoryInterface::ATTRIBUTES);
		}

		$this->customAttributesCodes[] = 'save_rewrites_history';

		return $this->customAttributesCodes;
	    }

would also destroy the purpose of modularity

Reply:
@swnsma  I am tempted to add a public method getControlAttributesCodes() to 
app/code/Magento/Catalog/Model/Category.php:302 which I then could plug-in overwrite in the UrlRewriteModule

But this would be a B/C break, right? But seems like the cleanest solution

Reply:
Or like it's done for addresses by injecting an attributeList Interface ... \Magento\Quote\Model\Quote\Address\CustomAttributeListInterface

Reply:
I of course could plug into all find of functions like \Magento\Framework\Model\AbstractExtensibleModel::getCustomAttributes and \Magento\Framework\Model\AbstractExtensibleModel::getCustomAttribute which implicitly call \Magento\Framework\Model\AbstractExtensibleModel::getCustomAttributesCodes and call getCustomAttributeCodes by myself and add the save_rewrites_history attribute, that would work without any other subsential changes, but is a would be a huge mess architecture-wise

Reply:
Or before adding a new public function, could we just make \Magento\Catalog\Model\Category::getCustomAttributesCodes public, so that a plugin could be used?

Reply:
Oh, I really would like to avoid such complications.
In such a case, it will be nice to have a mechanism for extension of Custom Attributes with not only EAV attributes but also with some configuration (more XML!).
But we already have Extension Attributes, which, could be used for this.

I prefer to look into two possible ways:
1) A short one - look into how we are saving category. 
Probably, we need to just copy 'all other' fields from category data as it works for the product: 
https://github.com/magento/magento2/blob/3008324fc7e90b01fd2b25be789aacaf950f6cde/app/code/Magento/Catalog/Model/ProductRepository.php#L538
2) A long one but more solid - look if we can replace usage of a custom attribute for extension attributes.
 But it still is required to think about possible BC breaks (at least for products, where it works previously).

Reply:
Thanks for your feedback @swnsma  - I choose the blue pill (1.) for now. Let's see what the full test suite say.
I might still be BC breaking in a way, that some module does not expect all data to be copied in the repository's save, but it is very unlikely.

Actually I am not sure how repositories are supposed to work: Whether they should the attributes to the ones clearly specified (like EAV attributes) or not.

Option 2. would be great to implement, because it also saves us from all the hassle in InputParamsResolver, but I just don't have the resources to do that now -- at least we would have a integration test now if this get's merged :)

In the end of the day, I am not even sure if this setData() approach is nice to trigger the URL key regeneration - maybe somebody has even some better idea here.

Reply:
I also do not really like the idea to pass all data with `setData()`. It also a hack but, at least, it does not provide modularity issue :sweat_smile:

**_But, in general, it will be nice to ask someone else for additional ideas._**

Reply:
I would prefer option 1 if it is enough to solve the issue.
Looks like the safest and the easiest way.

Reply:
That is like it is currently implemented in the PR... Can you take over and fix the tests? Thanks in advance